Fix Iceberg metadata listing failure when materialized view dropped

findepi opened the pull request:
Fix `SHOW TABLES` and `information_schema.tables` read failure when
materialized view is concurrently dropped

Fixes https://github.com/trinodb/trino/issues/10976

findepi commented on plugin/trino-sqlserver/src/test/java/io/trino/plugin/sqlserver/BaseSqlServerConnectorTest.java:
```diff
@@ -134,6 +135,20 @@ public void testReadFromView()
+                        super.testReadMetadataWithRelationsConcurrentModifications();
+                    }
+                })
+                // TODO https://github.com/trinodb/trino/issues/10846: shouldn't fail
```
https://github.com/trinodb/trino/issues/10846

findepi commented:
stress test green https://github.com/trinodb/trino/runs/4984192609?check_suite_focus=true except https://github.com/trinodb/trino/issues/9658 
dropping stress test commit

alexjo2144 commented on plugin/trino-sqlserver/src/test/java/io/trino/plugin/sqlserver/BaseSqlServerConnectorTest.java:
```diff
@@ -134,6 +136,21 @@ public void testReadFromView()
+                })
+                // TODO https://github.com/trinodb/trino/issues/10846: shouldn't fail
+                .hasMessageMatching("(?s).*(No task completed before timeout|was deadlocked on lock resources with another process and has been chosen as the deadlock victim).*");
+        throw new SkipException("to be fixed");
```
Did you mean to have some assertions and also a SkipException? Seems like you want one or the other

findepi replied:
was intentional

alexjo2144 reviewed (commented):
Neat test

findepi commented:
(rebased)

findepi commented:
(rebased)

findepi commented on plugin/trino-phoenix/src/test/java/io/trino/plugin/phoenix/TestPhoenixConnectorTest.java:
```diff
@@ -179,7 +179,7 @@ protected boolean isColumnNameRejected(Exception exception, String columnName, b
     public void testDataMappingSmokeTest(AbstractTestDistributedQueries.DataMappingTestSetup dataMappingTestSetup)
     {
-        // TODO enable the test
+        // TODO (https://github.com/trinodb/trino/issues/10905) enable the test
```
https://github.com/trinodb/trino/issues/10905

findepi commented on plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java:
```diff
@@ -196,7 +196,7 @@ protected boolean isColumnNameRejected(Exception exception, String columnName, b
     public void testDataMappingSmokeTest(DataMappingTestSetup dataMappingTestSetup)
     {
-        // TODO enable the test
+        // TODO (https://github.com/trinodb/trino/issues/10905) enable the test
```
https://github.com/trinodb/trino/issues/10905

findepi commented on plugin/trino-phoenix/src/test/java/io/trino/plugin/phoenix/TestPhoenixConnectorTest.java:
```diff
@@ -205,6 +205,15 @@ public void testShowCreateTable()
+    public void testReadMetadataWithRelationsConcurrentModifications()
+    {
+        assertThatThrownBy(super::testReadMetadataWithRelationsConcurrentModifications)
+                // TODO (https://github.com/trinodb/trino/issues/10904): shouldn't fail
```
https://github.com/trinodb/trino/issues/10904

findepi commented on plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java:
```diff
@@ -222,6 +222,15 @@ public void testShowCreateTable()
+    public void testReadMetadataWithRelationsConcurrentModifications()
+    {
+        assertThatThrownBy(super::testReadMetadataWithRelationsConcurrentModifications)
+                // TODO (https://github.com/trinodb/trino/issues/10904): shouldn't fail
```
https://github.com/trinodb/trino/issues/10904

findepi commented on plugin/trino-sqlserver/src/test/java/io/trino/plugin/sqlserver/BaseSqlServerConnectorTest.java:
```diff
@@ -134,6 +136,25 @@ public void testReadFromView()
+                        super.testReadMetadataWithRelationsConcurrentModifications();
+                    }
+                })
+                // TODO (https://github.com/trinodb/trino/issues/10846): shouldn't fail
```
https://github.com/trinodb/trino/issues/10846

findepi commented:
@hashhar can you ptal?

hashhar commented:
Iceberg tests are still failing:

```
2022-02-02T15:20:43.6658086Z [ERROR] Tests run: 756, Failures: 2, Errors: 0, Skipped: 4, Time elapsed: 1,439.977 s <<< FAILURE! - in TestSuite
2022-02-02T15:20:43.6659855Z [ERROR] io.trino.plugin.iceberg.TestIcebergParquetConnectorTest.testReadMetadataWithRelationsConcurrentModifications  Time elapsed: 112.863 s  <<< FAILURE!
2022-02-02T15:20:43.6661789Z java.util.concurrent.ExecutionException: java.lang.RuntimeException: Failed to open input stream for file: /tmp/TrinoTest9485126074854712339/iceberg_data/tpch/st_e61e57edaa264ae9be6b6a786422ae2c/metadata/00000-1599395c-7c7e-4ef3-9574-3c599b03f22b.metadata.json
2022-02-02T15:20:43.6662562Z 	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
2022-02-02T15:20:43.6663018Z 	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
2022-02-02T15:20:43.6663661Z 	at io.trino.testing.BaseConnectorTest.testReadMetadataWithRelationsConcurrentModifications(BaseConnectorTest.java:955)
2022-02-02T15:20:43.6664511Z 	at io.trino.testing.BaseConnectorTest.testReadMetadataWithRelationsConcurrentModifications(BaseConnectorTest.java:897)
2022-02-02T15:20:43.6665205Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2022-02-02T15:20:43.6665981Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2022-02-02T15:20:43.6666788Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2022-02-02T15:20:43.6667318Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2022-02-02T15:20:43.6667825Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2022-02-02T15:20:43.6668374Z 	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
2022-02-02T15:20:43.6668893Z 	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
2022-02-02T15:20:43.6669390Z 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
2022-02-02T15:20:43.6669847Z 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2022-02-02T15:20:43.6670494Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2022-02-02T15:20:43.6671076Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2022-02-02T15:20:43.6671508Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2022-02-02T15:20:43.6672382Z Caused by: java.lang.RuntimeException: Failed to open input stream for file: /tmp/TrinoTest9485126074854712339/iceberg_data/tpch/st_e61e57edaa264ae9be6b6a786422ae2c/metadata/00000-1599395c-7c7e-4ef3-9574-3c599b03f22b.metadata.json
2022-02-02T15:20:43.6673067Z 	at io.trino.testing.AbstractTestingTrinoClient.execute(AbstractTestingTrinoClient.java:122)
2022-02-02T15:20:43.6673621Z 	at io.trino.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:505)
2022-02-02T15:20:43.6674213Z 	at io.trino.testing.AbstractTestQueryFramework.computeActual(AbstractTestQueryFramework.java:139)
2022-02-02T15:20:43.6674854Z 	at io.trino.testing.AbstractTestQueryFramework.computeActual(AbstractTestQueryFramework.java:134)
2022-02-02T15:20:43.6675378Z 	at io.trino.testing.BaseConnectorTest$1.call(BaseConnectorTest.java:977)
2022-02-02T15:20:43.6675829Z 	at io.trino.testing.BaseConnectorTest$1.call(BaseConnectorTest.java:971)
2022-02-02T15:20:43.6676276Z 	at io.trino.testing.BaseConnectorTest$3.call(BaseConnectorTest.java:1045)
2022-02-02T15:20:43.6676719Z 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2022-02-02T15:20:43.6677013Z 	... 5 more
2022-02-02T15:20:43.6677455Z 	Suppressed: java.lang.Exception: SQL: SELECT * FROM system.metadata.table_comments WHERE catalog_name = CURRENT_CATALOG AND schema_name = CURRENT_SCHEMA
2022-02-02T15:20:43.6678032Z 		at io.trino.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:508)
2022-02-02T15:20:43.6678393Z 		... 11 more
2022-02-02T15:20:43.6678828Z 	Suppressed: java.lang.Exception: Task: Query(SELECT * FROM system.metadata.table_comments WHERE catalog_name = CURRENT_CATALOG AND schema_name = CURRENT_SCHEMA)
2022-02-02T15:20:43.6679371Z 		at io.trino.testing.BaseConnectorTest$3.call(BaseConnectorTest.java:1048)
2022-02-02T15:20:43.6679962Z 		... 6 more
2022-02-02T15:20:43.6680823Z Caused by: org.apache.iceberg.exceptions.NotFoundException: Failed to open input stream for file: /tmp/TrinoTest9485126074854712339/iceberg_data/tpch/st_e61e57edaa264ae9be6b6a786422ae2c/metadata/00000-1599395c-7c7e-4ef3-9574-3c599b03f22b.metadata.json
2022-02-02T15:20:43.6681500Z 	at org.apache.iceberg.hadoop.HadoopInputFile.newStream(HadoopInputFile.java:177)
2022-02-02T15:20:43.6682070Z 	at io.trino.plugin.hive.authentication.NoHdfsAuthentication.doAs(NoHdfsAuthentication.java:25)
2022-02-02T15:20:43.6682593Z 	at io.trino.plugin.hive.HdfsEnvironment.doAs(HdfsEnvironment.java:97)
2022-02-02T15:20:43.6683061Z 	at io.trino.plugin.iceberg.HdfsInputFile.newStream(HdfsInputFile.java:62)
2022-02-02T15:20:43.6683541Z 	at org.apache.iceberg.TableMetadataParser.read(TableMetadataParser.java:257)
2022-02-02T15:20:43.6684040Z 	at org.apache.iceberg.TableMetadataParser.read(TableMetadataParser.java:248)
2022-02-02T15:20:43.6684738Z 	at io.trino.plugin.iceberg.catalog.AbstractMetastoreTableOperations.lambda$refreshFromMetadataLocation$4(AbstractMetastoreTableOperations.java:296)
2022-02-02T15:20:43.6685429Z 	at org.apache.iceberg.util.Tasks$Builder.runTaskWithRetry(Tasks.java:405)
2022-02-02T15:20:43.6685855Z 	at org.apache.iceberg.util.Tasks$Builder.runSingleThreaded(Tasks.java:214)
2022-02-02T15:20:43.6686267Z 	at org.apache.iceberg.util.Tasks$Builder.run(Tasks.java:198)
2022-02-02T15:20:43.6686644Z 	at org.apache.iceberg.util.Tasks$Builder.run(Tasks.java:190)
2022-02-02T15:20:43.6687322Z 	at io.trino.plugin.iceberg.catalog.AbstractMetastoreTableOperations.refreshFromMetadataLocation(AbstractMetastoreTableOperations.java:295)
2022-02-02T15:20:43.6688160Z 	at io.trino.plugin.iceberg.catalog.AbstractMetastoreTableOperations.refresh(AbstractMetastoreTableOperations.java:160)
2022-02-02T15:20:43.6689016Z 	at io.trino.plugin.iceberg.catalog.AbstractMetastoreTableOperations.current(AbstractMetastoreTableOperations.java:132)
2022-02-02T15:20:43.6689675Z 	at io.trino.plugin.iceberg.TrinoHiveCatalog.lambda$loadTable$9(TrinoHiveCatalog.java:324)
2022-02-02T15:20:43.6690236Z 	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
2022-02-02T15:20:43.6690770Z 	at io.trino.plugin.iceberg.TrinoHiveCatalog.loadTable(TrinoHiveCatalog.java:322)
2022-02-02T15:20:43.6691328Z 	at io.trino.plugin.iceberg.TrinoHiveCatalog.doGetMaterializedView(TrinoHiveCatalog.java:641)
2022-02-02T15:20:43.6691913Z 	at io.trino.plugin.iceberg.TrinoHiveCatalog.lambda$getMaterializedView$23(TrinoHiveCatalog.java:612)
2022-02-02T15:20:43.6692397Z 	at net.jodah.failsafe.Functions.lambda$get$0(Functions.java:48)
2022-02-02T15:20:43.6692846Z 	at net.jodah.failsafe.RetryPolicyExecutor.lambda$supply$0(RetryPolicyExecutor.java:62)
2022-02-02T15:20:43.6693315Z 	at net.jodah.failsafe.Execution.executeSync(Execution.java:129)
2022-02-02T15:20:43.6693753Z 	at net.jodah.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
2022-02-02T15:20:43.6694374Z 	at net.jodah.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:67)
2022-02-02T15:20:43.6694916Z 	at io.trino.plugin.iceberg.TrinoHiveCatalog.getMaterializedView(TrinoHiveCatalog.java:612)
2022-02-02T15:20:43.6695499Z 	at io.trino.plugin.iceberg.IcebergMetadata.getMaterializedView(IcebergMetadata.java:1163)
2022-02-02T15:20:43.6696086Z 	at io.trino.spi.connector.ConnectorMetadata.getMaterializedViews(ConnectorMetadata.java:1269)
2022-02-02T15:20:43.6696818Z 	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.getMaterializedViews(ClassLoaderSafeConnectorMetadata.java:950)
2022-02-02T15:20:43.6697529Z 	at io.trino.metadata.MetadataManager.getMaterializedViews(MetadataManager.java:1355)
2022-02-02T15:20:43.6698072Z 	at io.trino.metadata.MetadataListing.getMaterializedViews(MetadataListing.java:153)
2022-02-02T15:20:43.6698637Z 	at io.trino.connector.system.TableCommentSystemTable.cursor(TableCommentSystemTable.java:108)
2022-02-02T15:20:43.6699207Z 	at io.trino.connector.system.SystemPageSourceProvider$1.cursor(SystemPageSourceProvider.java:129)
2022-02-02T15:20:43.6699846Z 	at io.trino.split.MappedRecordSet.cursor(MappedRecordSet.java:53)
2022-02-02T15:20:43.6700460Z 	at io.trino.spi.connector.RecordPageSource.<init>(RecordPageSource.java:37)
2022-02-02T15:20:43.6701026Z 	at io.trino.connector.system.SystemPageSourceProvider.createPageSource(SystemPageSourceProvider.java:108)
2022-02-02T15:20:43.6701594Z 	at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:68)
2022-02-02T15:20:43.6702165Z 	at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:268)
2022-02-02T15:20:43.6702776Z 	at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:196)
2022-02-02T15:20:43.6703308Z 	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:338)
2022-02-02T15:20:43.6703797Z 	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
2022-02-02T15:20:43.6704300Z 	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
2022-02-02T15:20:43.6704799Z 	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
2022-02-02T15:20:43.6705294Z 	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
2022-02-02T15:20:43.6705778Z 	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
2022-02-02T15:20:43.6706301Z 	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
2022-02-02T15:20:43.6706828Z 	at io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$3(WorkProcessorUtils.java:219)
2022-02-02T15:20:43.6707363Z 	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
2022-02-02T15:20:43.6707868Z 	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
2022-02-02T15:20:43.6708463Z 	at io.trino.operator.WorkProcessorUtils.lambda$finishWhen$4(WorkProcessorUtils.java:234)
2022-02-02T15:20:43.6709016Z 	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
2022-02-02T15:20:43.6709638Z 	at io.trino.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:150)
2022-02-02T15:20:43.6710161Z 	at io.trino.operator.Driver.processInternal(Driver.java:388)
2022-02-02T15:20:43.6711714Z 	at io.trino.operator.Driver.lambda$processFor$9(Driver.java:292)
2022-02-02T15:20:43.6714034Z 	at io.trino.operator.Driver.tryWithLock(Driver.java:682)
2022-02-02T15:20:43.6714413Z 	at io.trino.operator.Driver.processFor(Driver.java:285)
2022-02-02T15:20:43.6715048Z 	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1092)
2022-02-02T15:20:43.6715609Z 	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
2022-02-02T15:20:43.6716142Z 	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:488)
2022-02-02T15:20:43.6716542Z 	at io.trino.$gen.Trino_testversion____20220202_145654_6.run(Unknown Source)
2022-02-02T15:20:43.6716987Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2022-02-02T15:20:43.6717507Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2022-02-02T15:20:43.6717912Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2022-02-02T15:20:43.6718724Z Caused by: java.io.FileNotFoundException: File /tmp/TrinoTest9485126074854712339/iceberg_data/tpch/st_e61e57edaa264ae9be6b6a786422ae2c/metadata/00000-1599395c-7c7e-4ef3-9574-3c599b03f22b.metadata.json does not exist
2022-02-02T15:20:43.6719751Z 	at org.apache.hadoop.fs.RawLocalFileSystem.deprecatedGetFileStatus(RawLocalFileSystem.java:666)
2022-02-02T15:20:43.6720559Z 	at org.apache.hadoop.fs.RawLocalFileSystem.getFileLinkStatusInternal(RawLocalFileSystem.java:987)
2022-02-02T15:20:43.6721694Z 	at org.apache.hadoop.fs.RawLocalFileSystem.getFileStatus(RawLocalFileSystem.java:656)
2022-02-02T15:20:43.6722917Z 	at org.apache.hadoop.fs.FilterFileSystem.getFileStatus(FilterFileSystem.java:454)
2022-02-02T15:20:43.6723791Z 	at org.apache.hadoop.fs.ChecksumFileSystem$ChecksumFSInputChecker.<init>(ChecksumFileSystem.java:146)
2022-02-02T15:20:43.6724684Z 	at org.apache.hadoop.fs.ChecksumFileSystem.open(ChecksumFileSystem.java:347)
2022-02-02T15:20:43.6725402Z 	at io.trino.plugin.hive.fs.TrinoFileSystemCache$FileSystemWrapper.open(TrinoFileSystemCache.java:307)
2022-02-02T15:20:43.6726200Z 	at org.apache.hadoop.fs.FileSystem.open(FileSystem.java:899)
2022-02-02T15:20:43.6726640Z 	at org.apache.iceberg.hadoop.HadoopInputFile.newStream(HadoopInputFile.java:175)
2022-02-02T15:20:43.6726974Z 	... 61 more
2022-02-02T15:20:43.6727096Z 
2022-02-02T15:20:43.6727666Z [ERROR] io.trino.plugin.iceberg.TestIcebergOrcConnectorTest.testReadMetadataWithRelationsConcurrentModifications  Time elapsed: 6.114 s  <<< FAILURE!
2022-02-02T15:20:43.6729221Z java.util.concurrent.ExecutionException: java.lang.RuntimeException: Failed to open input stream for file: /tmp/TrinoTest14824247195884160678/iceberg_data/tpch/st_397b12f18dac4111821c5a90357d0048/metadata/00000-4a6b7679-13fe-40e8-8234-01de94f1cc41.metadata.json
2022-02-02T15:20:43.6730025Z 	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
2022-02-02T15:20:43.6730446Z 	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
2022-02-02T15:20:43.6731276Z 	at io.trino.testing.BaseConnectorTest.testReadMetadataWithRelationsConcurrentModifications(BaseConnectorTest.java:955)
2022-02-02T15:20:43.6732231Z 	at io.trino.testing.BaseConnectorTest.testReadMetadataWithRelationsConcurrentModifications(BaseConnectorTest.java:897)
2022-02-02T15:20:43.6733221Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2022-02-02T15:20:43.6734425Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2022-02-02T15:20:43.6735086Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2022-02-02T15:20:43.6735576Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2022-02-02T15:20:43.6736224Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2022-02-02T15:20:43.6736774Z 	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
2022-02-02T15:20:43.6737635Z 	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
2022-02-02T15:20:43.6738099Z 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
2022-02-02T15:20:43.6738542Z 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2022-02-02T15:20:43.6739033Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2022-02-02T15:20:43.6739730Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2022-02-02T15:20:43.6740466Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2022-02-02T15:20:43.6741592Z Caused by: java.lang.RuntimeException: Failed to open input stream for file: /tmp/TrinoTest14824247195884160678/iceberg_data/tpch/st_397b12f18dac4111821c5a90357d0048/metadata/00000-4a6b7679-13fe-40e8-8234-01de94f1cc41.metadata.json
2022-02-02T15:20:43.6742260Z 	at io.trino.testing.AbstractTestingTrinoClient.execute(AbstractTestingTrinoClient.java:122)
2022-02-02T15:20:43.6743013Z 	at io.trino.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:505)
2022-02-02T15:20:43.6743943Z 	at io.trino.testing.AbstractTestQueryFramework.computeActual(AbstractTestQueryFramework.java:139)
2022-02-02T15:20:43.6744756Z 	at io.trino.testing.AbstractTestQueryFramework.computeActual(AbstractTestQueryFramework.java:134)
2022-02-02T15:20:43.6745684Z 	at io.trino.testing.BaseConnectorTest$1.call(BaseConnectorTest.java:986)
2022-02-02T15:20:43.6746745Z 	at io.trino.testing.BaseConnectorTest$1.call(BaseConnectorTest.java:971)
2022-02-02T15:20:43.6747233Z 	at io.trino.testing.BaseConnectorTest$3.call(BaseConnectorTest.java:1045)
2022-02-02T15:20:43.6748230Z 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2022-02-02T15:20:43.6748576Z 	... 5 more
2022-02-02T15:20:43.6749382Z 	Suppressed: java.lang.Exception: SQL: SELECT * FROM system.metadata.materialized_views WHERE catalog_name = CURRENT_CATALOG AND schema_name = CURRENT_SCHEMA
2022-02-02T15:20:43.6750711Z 		at io.trino.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:508)
2022-02-02T15:20:43.6751303Z 		... 11 more
2022-02-02T15:20:43.6751783Z 	Suppressed: java.lang.Exception: Task: Query(SELECT * FROM system.metadata.materialized_views WHERE catalog_name = CURRENT_CATALOG AND schema_name = CURRENT_SCHEMA)
2022-02-02T15:20:43.6752768Z 		at io.trino.testing.BaseConnectorTest$3.call(BaseConnectorTest.java:1048)
2022-02-02T15:20:43.6753104Z 		... 6 more
2022-02-02T15:20:43.6754136Z Caused by: org.apache.iceberg.exceptions.NotFoundException: Failed to open input stream for file: /tmp/TrinoTest14824247195884160678/iceberg_data/tpch/st_397b12f18dac4111821c5a90357d0048/metadata/00000-4a6b7679-13fe-40e8-8234-01de94f1cc41.metadata.json
2022-02-02T15:20:43.6754929Z 	at org.apache.iceberg.hadoop.HadoopInputFile.newStream(HadoopInputFile.java:177)
2022-02-02T15:20:43.6756048Z 	at io.trino.plugin.hive.authentication.NoHdfsAuthentication.doAs(NoHdfsAuthentication.java:25)
2022-02-02T15:20:43.6756575Z 	at io.trino.plugin.hive.HdfsEnvironment.doAs(HdfsEnvironment.java:97)
2022-02-02T15:20:43.6757057Z 	at io.trino.plugin.iceberg.HdfsInputFile.newStream(HdfsInputFile.java:62)
2022-02-02T15:20:43.6758763Z 	at org.apache.iceberg.TableMetadataParser.read(TableMetadataParser.java:257)
2022-02-02T15:20:43.6759376Z 	at org.apache.iceberg.TableMetadataParser.read(TableMetadataParser.java:248)
2022-02-02T15:20:43.6945619Z 	at io.trino.plugin.iceberg.catalog.AbstractMetastoreTableOperations.lambda$refreshFromMetadataLocation$4(AbstractMetastoreTableOperations.java:296)
2022-02-02T15:20:43.6946728Z 	at org.apache.iceberg.util.Tasks$Builder.runTaskWithRetry(Tasks.java:405)
2022-02-02T15:20:43.6947222Z 	at org.apache.iceberg.util.Tasks$Builder.runSingleThreaded(Tasks.java:214)
2022-02-02T15:20:43.6947867Z 	at org.apache.iceberg.util.Tasks$Builder.run(Tasks.java:198)
2022-02-02T15:20:43.6948446Z 	at org.apache.iceberg.util.Tasks$Builder.run(Tasks.java:190)
2022-02-02T15:20:43.6949183Z 	at io.trino.plugin.iceberg.catalog.AbstractMetastoreTableOperations.refreshFromMetadataLocation(AbstractMetastoreTableOperations.java:295)
2022-02-02T15:20:43.6950138Z 	at io.trino.plugin.iceberg.catalog.AbstractMetastoreTableOperations.refresh(AbstractMetastoreTableOperations.java:160)
2022-02-02T15:20:43.6950989Z 	at io.trino.plugin.iceberg.catalog.AbstractMetastoreTableOperations.current(AbstractMetastoreTableOperations.java:132)
2022-02-02T15:20:43.6951703Z 	at io.trino.plugin.iceberg.TrinoHiveCatalog.lambda$loadTable$9(TrinoHiveCatalog.java:324)
2022-02-02T15:20:43.6952325Z 	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
2022-02-02T15:20:43.6952919Z 	at io.trino.plugin.iceberg.TrinoHiveCatalog.loadTable(TrinoHiveCatalog.java:322)
2022-02-02T15:20:43.6953538Z 	at io.trino.plugin.iceberg.TrinoHiveCatalog.doGetMaterializedView(TrinoHiveCatalog.java:641)
2022-02-02T15:20:43.6954184Z 	at io.trino.plugin.iceberg.TrinoHiveCatalog.lambda$getMaterializedView$23(TrinoHiveCatalog.java:612)
2022-02-02T15:20:43.6954694Z 	at net.jodah.failsafe.Functions.lambda$get$0(Functions.java:48)
2022-02-02T15:20:43.6955211Z 	at net.jodah.failsafe.RetryPolicyExecutor.lambda$supply$0(RetryPolicyExecutor.java:62)
2022-02-02T15:20:43.6955913Z 	at net.jodah.failsafe.Execution.executeSync(Execution.java:129)
2022-02-02T15:20:43.6956419Z 	at net.jodah.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
2022-02-02T15:20:43.6957066Z 	at net.jodah.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:67)
2022-02-02T15:20:43.6957643Z 	at io.trino.plugin.iceberg.TrinoHiveCatalog.getMaterializedView(TrinoHiveCatalog.java:612)
2022-02-02T15:20:43.6958433Z 	at io.trino.plugin.iceberg.IcebergMetadata.getMaterializedView(IcebergMetadata.java:1163)
2022-02-02T15:20:43.6959813Z 	at io.trino.spi.connector.ConnectorMetadata.getMaterializedViews(ConnectorMetadata.java:1269)
2022-02-02T15:20:43.6961743Z 	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.getMaterializedViews(ClassLoaderSafeConnectorMetadata.java:950)
2022-02-02T15:20:43.6962573Z 	at io.trino.metadata.MetadataManager.getMaterializedViews(MetadataManager.java:1355)
2022-02-02T15:20:43.6963197Z 	at io.trino.metadata.MetadataListing.getMaterializedViews(MetadataListing.java:153)
2022-02-02T15:20:43.6963896Z 	at io.trino.connector.system.MaterializedViewSystemTable.lambda$cursor$1(MaterializedViewSystemTable.java:103)
2022-02-02T15:20:43.6964523Z 	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422)
2022-02-02T15:20:43.6965168Z 	at com.google.common.collect.RegularImmutableSortedSet.forEach(RegularImmutableSortedSet.java:88)
2022-02-02T15:20:43.6965898Z 	at io.trino.connector.system.MaterializedViewSystemTable.cursor(MaterializedViewSystemTable.java:100)
2022-02-02T15:20:43.6966600Z 	at io.trino.connector.system.SystemPageSourceProvider$1.cursor(SystemPageSourceProvider.java:129)
2022-02-02T15:20:43.6967159Z 	at io.trino.split.MappedRecordSet.cursor(MappedRecordSet.java:53)
2022-02-02T15:20:43.6967686Z 	at io.trino.spi.connector.RecordPageSource.<init>(RecordPageSource.java:37)
2022-02-02T15:20:43.6968334Z 	at io.trino.connector.system.SystemPageSourceProvider.createPageSource(SystemPageSourceProvider.java:108)
2022-02-02T15:20:43.6969022Z 	at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:68)
2022-02-02T15:20:43.6969698Z 	at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:268)
2022-02-02T15:20:43.6970760Z 	at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:196)
2022-02-02T15:20:43.6971381Z 	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:338)
2022-02-02T15:20:43.6971958Z 	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
2022-02-02T15:20:43.6972519Z 	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
2022-02-02T15:20:43.6973086Z 	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
2022-02-02T15:20:43.6973632Z 	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
2022-02-02T15:20:43.6974364Z 	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
2022-02-02T15:20:43.7015310Z 	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
2022-02-02T15:20:43.7015912Z 	at io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$3(WorkProcessorUtils.java:219)
2022-02-02T15:20:43.7016501Z 	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
2022-02-02T15:20:43.7017088Z 	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
2022-02-02T15:20:43.7017650Z 	at io.trino.operator.WorkProcessorUtils.lambda$finishWhen$4(WorkProcessorUtils.java:234)
2022-02-02T15:20:43.7018222Z 	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
2022-02-02T15:20:43.7018891Z 	at io.trino.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:150)
2022-02-02T15:20:43.7019491Z 	at io.trino.operator.Driver.processInternal(Driver.java:388)
2022-02-02T15:20:43.7019919Z 	at io.trino.operator.Driver.lambda$processFor$9(Driver.java:292)
2022-02-02T15:20:43.7020319Z 	at io.trino.operator.Driver.tryWithLock(Driver.java:682)
2022-02-02T15:20:43.7020731Z 	at io.trino.operator.Driver.processFor(Driver.java:285)
2022-02-02T15:20:43.7021231Z 	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1092)
2022-02-02T15:20:43.7021847Z 	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
2022-02-02T15:20:43.7023459Z 	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:488)
2022-02-02T15:20:43.7024279Z 	at io.trino.$gen.Trino_testversion____20220202_145654_5.run(Unknown Source)
2022-02-02T15:20:43.7025101Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2022-02-02T15:20:43.7025672Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2022-02-02T15:20:43.7026099Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2022-02-02T15:20:43.7027057Z Caused by: java.io.FileNotFoundException: File /tmp/TrinoTest14824247195884160678/iceberg_data/tpch/st_397b12f18dac4111821c5a90357d0048/metadata/00000-4a6b7679-13fe-40e8-8234-01de94f1cc41.metadata.json does not exist
2022-02-02T15:20:43.7027773Z 	at org.apache.hadoop.fs.RawLocalFileSystem.deprecatedGetFileStatus(RawLocalFileSystem.java:666)
2022-02-02T15:20:43.7028456Z 	at org.apache.hadoop.fs.RawLocalFileSystem.getFileLinkStatusInternal(RawLocalFileSystem.java:987)
2022-02-02T15:20:43.7029077Z 	at org.apache.hadoop.fs.RawLocalFileSystem.getFileStatus(RawLocalFileSystem.java:656)
2022-02-02T15:20:43.7029656Z 	at org.apache.hadoop.fs.FilterFileSystem.getFileStatus(FilterFileSystem.java:454)
2022-02-02T15:20:43.7030241Z 	at org.apache.hadoop.fs.ChecksumFileSystem$ChecksumFSInputChecker.<init>(ChecksumFileSystem.java:146)
2022-02-02T15:20:43.7030795Z 	at org.apache.hadoop.fs.ChecksumFileSystem.open(ChecksumFileSystem.java:347)
2022-02-02T15:20:43.7031353Z 	at io.trino.plugin.hive.fs.TrinoFileSystemCache$FileSystemWrapper.open(TrinoFileSystemCache.java:307)
2022-02-02T15:20:43.7031868Z 	at org.apache.hadoop.fs.FileSystem.open(FileSystem.java:899)
2022-02-02T15:20:43.7032367Z 	at org.apache.iceberg.hadoop.HadoopInputFile.newStream(HadoopInputFile.java:175)
2022-02-02T15:20:43.7032865Z 	... 64 more
2022-02-02T15:20:43.7033004Z 
2022-02-02T15:20:43.8090294Z [INFO] 
2022-02-02T15:20:43.8090791Z [INFO] Results:
2022-02-02T15:20:43.8104164Z [INFO] 
2022-02-02T15:20:43.8107713Z [ERROR] Failures: 
2022-02-02T15:20:43.8109709Z [ERROR]   TestIcebergOrcConnectorTest>BaseConnectorTest.testReadMetadataWithRelationsConcurrentModifications:897->BaseConnectorTest.testReadMetadataWithRelationsConcurrentModifications:955 » Execution
2022-02-02T15:20:43.8111627Z [ERROR]   TestIcebergParquetConnectorTest>BaseConnectorTest.testReadMetadataWithRelationsConcurrentModifications:897->BaseConnectorTest.testReadMetadataWithRelationsConcurrentModifications:955 » Execution
```

hashhar commented on testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:
```diff
@@ -866,6 +878,181 @@ protected void checkInformationSchemaViewsForMaterializedView(String schemaName,
     }
 
+    /**
+     * Test that reading table, column metadata, like {@code SHOW TABLES} or reading from {@code information_schema.views}
+     * does not fail when relations are concurrently created or dropped.
```
Generally speaking do we want to "fix" this in all connectors? It's possible that the underlying system itself exhibits this behavior and we can do nothing except retries.

findepi replied:
Doing retries could be good way of "fixing".
If the underlying system is inherently problematic, disabling the test is OK.

hashhar commented on testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:
```diff
@@ -866,6 +878,181 @@ protected void checkInformationSchemaViewsForMaterializedView(String schemaName,
+
+        int readIterations = 10;
+        // generous timeout as this is a generic test; typically should be faster
+        int testTimeoutSeconds = 150;
```
This isn't timeout for the entire test - it's just for the test setup (i.e. first create + drop starts running). Can you rename to clarify?

findepi replied:
No, it's used after setup too, see
```
completionService.poll(testTimeoutSeconds - testWatch.elapsed(SECONDS), SECONDS)
```

hashhar replied:
Good point. Upper bound for test (at-least the tasks) is decided by this value.

hashhar reviewed (approved):
The test itself looks fine to me.

I didn't look at the Iceberg fixes in depth - I'm not aware of all failure modes. cc: @alexjo2144 for that part.

LGTM % Iceberg fix review.

findepi commented:
(rebased to resolve conflict in TestPhoenixConnectorTest)

findepi commented:
> Iceberg tests are still failing:
> 
> ```
> 2022-02-02T15:20:43.6658086Z [ERROR] Tests run: 756, Failures: 2, Errors: 0, Skipped: 4, Time elapsed: 1,439.977 s <<< FAILURE! - in TestSuite
> 2022-02-02T15:20:43.6659855Z [ERROR] io.trino.plugin.iceberg.TestIcebergParquetConnectorTest.testReadMetadataWithRelationsConcurrentModifications  Time elapsed: 112.863 s  <<< FAILURE!
> 2022-02-02T15:20:43.6661789Z java.util.concurrent.ExecutionException: java.lang.RuntimeException: Failed to open input stream for file: /tmp/TrinoTest9485126074854712339/iceberg_data/tpch/st_e61e57edaa264ae9be6b6a786422ae2c/metadata/00000-1599395c-7c7e-4ef3-9574-3c599b03f22b.metadata.json
> 2022-02-02T15:20:43.6662562Z 	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
> ```

thanks @hashhar . sadly, i wasn't able to reproduce the problem. apparent interleaving on CI happens to be different.

findepi commented:
was green (https://github.com/trinodb/trino/actions/runs/1800525457)
rebased (conflicts)

findepi commented:
was green (https://github.com/trinodb/trino/actions/runs/1800525457)
rebased (conflicts)

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -1150,6 +1157,22 @@ public ConnectorInsertTableHandle beginRefreshMaterializedView(ConnectorSession
     }
 
+    @Override
+    public Map<SchemaTableName, ConnectorMaterializedViewDefinition> getMaterializedViews(ConnectorSession session, Optional<String> schemaName)
```
Should this change just go into the default implementation in `ConnectorMetadata`?

findepi replied:
Iceberg has a challenge because it keeps state between metastore and filesystem, and cannot rely on observing consistent state. I would hope most connectors don't have problems like that. Or, may be able to better exception-control this (here i catch `RuntimeException` which is too broad in a general case).

alexjo2144 reviewed (approved):
Couple clarifying questions but looks good to me

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java:
```diff
@@ -598,6 +603,22 @@ public void dropMaterializedView(ConnectorSession session, SchemaTableName schem
+            return Failsafe.with(new RetryPolicy<>()
+                            .withMaxAttempts(10)
+                            .withBackoff(100, 5_000, ChronoUnit.MILLIS)
+                            .withMaxDuration(Duration.ofSeconds(30))
```
correct, because we invalidate

findepi commented:
rebased on top of https://github.com/trinodb/trino/pull/10977 to make sure it gets reverted when the problem https://github.com/trinodb/trino/issues/10976 is fixed

findepi commented:
rebased on current master, so that stacktraces from CI match the code.

findepi commented:
resolved (snd squashed fixups) to resolve conflicts

findepi commented:
rebased after https://github.com/trinodb/trino/pull/11134

findepi commented:
Pin-pointing the problem took me a bit, but it seems it's hopefully all good now.
Removed all the TMP commits and rebased.

@hashhar @losipiuk let me know if you want to take another look, or i should just merge it.

findepi commented:
```
58 successful and 2 failing checks
ci / test (:trino-kudu) (pull_request) Failing after 34m — test (:trino-kudu)
ci / test (:trino-phoenix,:trino-phoenix5) (pull_request) Failing after 23m — test (:trino-phoenix,:trino-phoenix5)
```

- kudu cannot create table without partitioning (known issue)
- phoenix -> https://github.com/trinodb/trino/issues/10904#issuecomment-1052286822

will disable

hashhar reviewed (approved):
LGTM % someone more familiar with Iceberg to look at "Fix Iceberg metadata listing failure when materialized view dropped". The fix looks ok to me but I don't know all failure modes in Iceberg connector. cc: @alexjo2144

findepi commented:
Conflict, will rebase to resolve.

findepi commented:
CI `ci / test (plugin/trino-hive, test-failure-recovery)` - https://github.com/trinodb/trino/issues/11196 / https://github.com/trinodb/trino/issues/10631
